src/esphome/output/sm16716_output_component.cpp

```diff
@@ -0,0 +1,126 @@
+
+#include "esphome/defines.h"
+
+#ifdef USE_SM16716_OUTPUT
+
+#include "esphome/log.h"
+#include "esphome/output/sm16716_output_component.h"
+
+ESPHOME_NAMESPACE_BEGIN
+
+namespace output {
+
+static const char *TAG = "output.sm16716";
+
+SM16716OutputComponent::SM16716OutputComponent(GPIOPin *pin_data, GPIOPin *pin_clock, uint8_t num_channels,
+                                               uint8_t num_chips, bool update)
+    : pin_data_(pin_data), pin_clock_(pin_clock), num_channels_(num_channels), num_chips_(num_chips), update_(update) {}
+
+void SM16716OutputComponent::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up SM16716OutputComponent...");
+  this->pin_data_->setup();
+  this->pin_data_->digital_write(false);
+  this->pin_clock_->setup();
+  this->pin_clock_->digital_write(false);
+  this->pwm_amounts_.resize(this->num_channels_, 0);
+}
+void SM16716OutputComponent::dump_config() {
+  ESP_LOGCONFIG(TAG, "SM16716:");
+  LOG_PIN("  DATA Pin: ", this->pin_data_);
+  LOG_PIN("  CLOCK Pin: ", this->pin_clock_);
+  ESP_LOGCONFIG(TAG, "  Total number of channels: %u", this->num_channels_);
+  ESP_LOGCONFIG(TAG, "  Number of chips: %u", this->num_chips_);
+}
+
+void SM16716OutputComponent::loop() {
+  if (!this->update_) {
+    return;
+  }
+
+  for (uint8_t i = 0; i < 50; i++) {
+    this->write_bit_(false);
+  }
+
+  // send 25 bits (1 start bit plus 24 data bits) for each chip
+  uint8_t index = 0;
+  while (index < this->num_channels_) {
+    // send a start bit initially and after every 3 channels
+    if (index % 3 == 0) {
+      this->write_bit_(true);
+    }
+
+    this->write_byte_(this->pwm_amounts_[index]);
+    index++;
+  }
+
+  // send a blank 25 bits to signal the end
+  this->write_bit_(false);
+  this->write_byte_(0);
+  this->write_byte_(0);
+  this->write_byte_(0);
+
+  this->update_ = false;
+}
+
+SM16716OutputComponent::Channel *SM16716OutputComponent::create_channel(uint8_t channel,
+                                                                        PowerSupplyComponent *power_supply,
+                                                                        float max_power) {
+  ESP_LOGV(TAG, "Getting channel %d...", channel);
+  auto *c = new Channel(this, channel);
+  c->set_power_supply(power_supply);
+  c->set_max_power(max_power);
+  return c;
+}
+
+float SM16716OutputComponent::get_setup_priority() const { return setup_priority::HARDWARE; }
+
+void SM16716OutputComponent::set_channel_value_(uint8_t channel, uint8_t value) {
+  ESP_LOGV(TAG, "set channels %u to %u", channel, value);
+  uint8_t index = this->num_channels_ - channel - 1;
+  if (this->pwm_amounts_[index] != value) {
+    this->update_ = true;
+  }
+  this->pwm_amounts_[index] = value;
+}
+
+void SM16716OutputComponent::write_bit_(bool value) {
+  this->pin_data_->digital_write(value);
+  this->pin_clock_->digital_write(true);
+  this->pin_clock_->digital_write(false);
```

Comment:
Best I can tell by sending the Chinese datasheet through a translator is that the chip can handle 30 MHz. This seems to be the interpretation of the [Tasmota implementation](https://github.com/arendst/Sonoff-Tasmota/blob/master/sonoff/xdrv_04_light.ino#L382) as well. Seems fine at least with the relatively slow 8266 in the Feit bulbs I'm testing with, but I suppose it could be a problem if someone hooked up a chip to a higher clocked ESP32. Is there any good examples of adjusting timing to account for the actual clock speed? I'm pretty new to this level of embedded programming and am mostly in the "copy and tweak" rather than fully original coding level. :)